src/iris/api.py

```diff
@@ -1825,6 +1910,87 @@ def on_post(self, req, resp):
         resp.set_header('Location', '/plans/%s' % plan_id)
 
 
+def count_fields_for_incidents(cursor, where_clause, fields, allowed_columns, extra_joins=''):
+    field_counts = {}
+    non_tag_fields = [f for f in fields if f != 'tags']
+
+    total_count = 0
+
+    # Consolidated non-tag query
+    if non_tag_fields:
+        union_queries = []
+        for field in non_tag_fields:
+            if field not in allowed_columns:
+                continue
+            expr = allowed_columns[field].split(' as ')[0].strip()
+            subquery = f"""
+                SELECT '{field}' AS field, {expr} AS value, COUNT(*) AS group_count
+                FROM incident
+                JOIN plan ON incident.plan_id = plan.id
+                JOIN application ON incident.application_id = application.id
+                LEFT OUTER JOIN target ON incident.owner_id = target.id
+                LEFT OUTER JOIN template_variable
+                    ON template_variable.application_id = application.id
+                    AND template_variable.title_variable = 1
+                {extra_joins}
+                WHERE {where_clause}
+                GROUP BY {expr}
+            """
+            union_queries.append(subquery)
+
+        final_query = " UNION ALL ".join(union_queries)
+        cursor.execute(final_query)
+        for row in cursor.fetchall():
+            field = row['field']
+            value = row['value']
+            count = row['group_count']
+            field_counts.setdefault(field, {})[value] = count
+        if field_counts:
+            total_count = sum(field_counts[next(iter(field_counts))].values())
```

Comment:
we can wrap these into separate functions since these seem to be getting pretty large with consolidations for non_tag_fields, tags, and not non_tag_fields. For myself that's not familiar with this, I'm constantly checking to see if any of the current variables are populated from the previous clause which separate functions can disambiguate.

Reply:
I refactored them into more modular functions for clarity